<!DOCTYPE html>
<html lang="pt-BR">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link rel="shortcut icon" href="../static/assets/icon.svg" type="image/x-icon" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/home.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/panEDTA.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/help.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/results.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/dark-theme.css') }}" />
    <title> EDTA </title>
</head>
<body>
	<div class="menuMobile">
		<img src="../static/assets/icon_menu.png" alt="menu icon" class="menuIcon">
	</div>

	<aside class="aside">
		<ul id="menu">
			<div class="logoedta home" id="Home">
				<img src="../static/assets/icon.svg" alt="Logo do edta">
			</div>
			<li class="li open home" id="Home">
				<img class="home-icon" src="../static/assets/home.svg">
				<span class="txt-link"> EDTA </span>
			</li>
			<li class="li" id="panEDTA">
				<img class="help-icon" src="../static/assets/dna.svg">
				<span class="txt-link"> panEDTA </span>
			</li>
			<li class="li" id="Results">
				<img class="help-icon" src="../static/assets/box.svg">
				<span class="txt-link"> Results </span>
			</li>
			<li class="li" id="Help">
				<img class="help-icon" src="../static/assets/QuestionCircle.svg">
				<span class="txt-link"> Help </span>
			</li>
		</ul>
	</aside>

	<div class="cloudImageWrapper">
		<img id="cloudImage" class="replace" src="../static/assets/CloudMoon.svg">
	</div>

	<main class="main sectionsPage">
		<header class="header">
			<section class="container-header">
				<img class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>

		<section class="container-main">
			<article class="description-screen">
				<p> Welcome to EDTAGUI. This version provides a graphical interface that streamlines your work and research in annotating transposable elements. The performance of the annotation process depends on your machine's specifications, so ensure it remains powered on during execution. For any questions or further information, please use the <b>'Help'</b> option available in the menu.</p>
			</article>
		</section>
	

		<section class="container-data-input">
			<div class="title-input">
				<p>Data input</p>
			</div>
			<div class="data-input">
				<section class="container-email">
					<p class="session-title">Email Adress</p>
					<input name=email id="email" class="email" placeholder="Enter email" type="email" onfocus="clearPlaceholder(this)"/> 
				</section>
				
				<section class="container-genome">
					<p class="session-title">Genome data <span class="required">*</span> </p>
					<div  class="input-genome">
						<div class="box-input">
							<input type="text" id="fileNameInput" placeholder="FASTA" class="input-text" readonly disabled>
		
							<label for="inputdata" class="file-input">
								<p class="name-btn">Browser</p>
								<input name="genome-file" type="file" class="input-data" accept=".fasta,.fa,.fna,.ffn,.faa,.frn" id="inputdata">
							</label>
						</div>
					</div>
				</section>
			</div>
		</section>

		<section class="container-action">
			<div class="title-add-features">
				<p>Additional features</p>
			</div>

			<div class="container-add-feature">
				<section>
					<p class="session-title">Species specification to identify TIR candidates </p>
					<div class="box-checkboxs-tir">
						<div class="container-checkbox-tir">
							<input type="checkbox" id="others" name="tir" value="others" class="checkbox-tir" checked onclick="toggleCheckbox('box-checkboxs-tir', this)">
							<label for="others">others</label>
						</div>
						<div class="container-checkbox-tir">
							<input type="checkbox" id="rice" name="tir" value="rice" class="checkbox-tir" onclick="toggleCheckbox('box-checkboxs-tir', this)">
							<label for="rice">rice</label>
						</div>
						<div class="container-checkbox-tir">
							<input type="checkbox" id="maize" name="tir" value="maize" class="checkbox-tir" onclick="toggleCheckbox('box-checkboxs-tir', this)">
							<label for="maize">maize</label>
						</div>
					</div>
				</section>
			
				<section>
					<p class="session-title">Steps to be executed </p>
					<div class="box-checkboxs-step">
						<div class="container-checkbox-step">
							<input type="checkbox" id="all" name="step" value="all" class="checkbox-step" checked onclick="toggleCheckbox('box-checkboxs-step', this); updateDirectoryInputState()">
							<label for="all">all</label>
						</div>
						<div class="container-checkbox-step">
							<input type="checkbox" id="filter" name="step" value="filter" class="checkbox-step" onclick="toggleCheckbox('box-checkboxs-step', this); updateDirectoryInputState()">
							<label for="filter">filter</label>
						</div>
						<div class="container-checkbox-step">
							<input type="checkbox" id="final" name="step" value="final" class="checkbox-step" onclick="toggleCheckbox('box-checkboxs-step', this); updateDirectoryInputState()">
							<label for="final">final</label>
						</div>
						<div class="container-checkbox-step">
							<input type="checkbox" id="anno" name="step" value="anno" class="checkbox-step" onclick="toggleCheckbox('box-checkboxs-step', this); updateDirectoryInputState()">
							<label for="anno">anno</label>
						</div>
						<div class="container-directory-name">
							<input name=directory_name id="directoryInput" class="directory_name" placeholder="Enter directory name" type="text" onfocus="clearPlaceholder(this)" readonly disabled/> 
						</div>
					</div>
				</section>
			</div>
			

		   <section class="container-switch">
				<table class="table-switch">
					<tr>
						<td><p class="session-title">Overwrite</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch1" name="overwrite"/>
								<label for="switch1">Toggle</label>
							</div>
						</td>
						<td><p id="status-text-switch1"> Deactivated </p></td>
					</tr>
					<tr>
						<td><p class="session-title">Sensitivity</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch2" name="sensitivity" checked/>
								<label for="switch2">Toggle</label>
							</div>
						</td>
						<td><p id="status-text-switch2"> Enable</p></td>
					</tr>
					<tr>
						<td><p class="session-title">Annotation</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch3" name="annotation"/>
								<label for="switch3">Toggle</label>
							</div>
						</td>
						<td><p id="status-text-switch3"> Deactivated</p></td>
					</tr>
				</table>

				<table class="table-switch">
					<tr>
						<td><p class="session-title">Evaluate</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch4" name="evaluate"/>
								<label for="switch4">Toggle</label>   
							</div>
						</td>
						<td><p id="status-text-switch4">Deactivated</p></td>
					</tr>
					<tr>
						<td><p class="session-title">Force</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch5" name="force"/>
								<label for="switch5">Toggle</label>
							</div>
						</td>
						<td><p id="status-text-switch5">Deactivated</p></td>
					</tr>
				</table>

		   </section>

			<div class="container-number-input">
				<section class="container-mutation-rate">
					<label for="mutation-rate" class="session-title"> Neutral mutation rate </label>
					<input type="text" id="mutation-rate" name="mutation_rate" placeholder="1.3e-8" title="Enter a value in scientific notation, for example, 1.3e-8"/>
				</section>
	
				<section class="container-maxdiv">
					<label for="maximun-divergence" class="session-title"> Maximum divergence </label>
					<input type="number" id="maximun-divergence" name="max_divergence" value="40" min="0" max="100" title="Enter a value from 0 to 100"/>
				</section>
	
				<section class="container-threads">
					<p class="session-title"> Threads </p>
					<input type="number" id="threads" name="threads" min="4" value="4" step="1">
				</section>
			</div>
			<span class="not-required">* For highly repetitive genomes, users are encouraged to adjust the "Maximum divergence" parameter </span>
		</section>

			<section class="container-action">
				<div class="title-add-features">
					<p>Additional input files</p>
					<span class="not-required"> (Field not required) </span>
				</div>

				<div class="container-add-input">
					<section class="container-genome">
						<div class="session-title">
							<p> Coding DNA Sequence (cds) </p>
						</div>
						<div class="box-input">
							<input type="text" id="cdsInput" name="cds_file" class="input-library" placeholder=".fasta .fa" readonly disabled>
							<label for="cds" class="file-input">
								<p class="name-btn">Browser</p>
								<input type="file" class="input-data" accept="fasta,.fa,.fna,.ffn,.faa,.frn" id="cds">
							</label>
						</div>
					</section>

					<section class="container-genome">
						<div class="session-title">
							<p> Curate library (curatedlib) </p>
						</div>
						<div class="box-input">
							<input type="text" id="curateLibInput" name="curate_lib" class="input-library"  readonly disabled>
							<label for="curatelib" class="file-input">
								<p class="name-btn">Browser</p>
								<input type="file" class="input-data" id="curatelib">
							</label>
						</div>
					</section>

					<section class="container-genome">
						<div class="session-title">
							<p>Exclusion of masked regions (exclude)</p>
						</div>
						<div class="box-input">
							<input type="text" id="maskedRegionsInput" name="masked_regions" class="input-library" placeholder=".bed" readonly disabled>
		
							<label for="exclude" class="file-input">
								<p class="name-btn disabled" id="browseExcludeButton">Browser</p>
								  <input type="file" class="input-data" accept=".bed" id="exclude">
							</label>
						</div>
					</section>
				</div>

				<div class="container-add-input">
					<section class="container-genome">
						<div class="session-title">
							<p>RepeatModeler library (rmlib)</p>
						</div>
						<div class="box-input">
							<input type="text" id="rmLibInput" name="rm_lib" class="input-library" readonly disabled>
		
							<label for="rmlib" class="file-input">
								<p class="name-btn">Browser</p>
								  <input type="file" class="input-data" id="rmlib">
							</label>
						</div>
					</section>

					<section class="container-genome">
						<div class="session-title">
							<p> RepeatMasker library (rmout)</p>
						</div>
						<div class="box-input">
							<input type="text" id="rmoutInput" name="rm_out" class="input-library" placeholder=".out" readonly disabled>
							<label for="rmout" class="file-input">
								<p class="name-btn disabled" id="browseButton">Browser</p>
								<input type="file" class="input-data" accept=".out" id="rmout" disabled>
							</label>
						</div>
					</section>
				</div>
			</section>
	
		<section class="container-submit">
			<button type="button" id="uploaddata" class="uploaddata"> Submit </button>
		</section>

		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>

	<main class="panEDTA">
		<header class="header">
			<section class="container-header">
				<img class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>
		<section class="container-panGenome">
			<div class="title-input">
				<p>panEDTA</p>
			</div>
			<section class="container-main">
				<article class="description-screen">
					<p>This is the serial version of panEDTA. Each genome will be annotated sequentially and then combined with the panEDTA functionality. Existing EDTA annotation of genomes (EDTA run with --anno 1) will be recognized and reused. A way to acclerate the pan-genome annotation is to execute EDTA annotation of each genomes separately and in parallel, then execute panEDTA to finish the remaining of the runs. You may want to save the GFF files and the sum file of the EDTA results of each genome because they will be overwritten by panEDTA. To help filtering out gene-related sequences, at least one CDS file is required. Please read <a href="https://github.com/oushujun/EDTA/wiki/Making-sense-of-EDTA-usage-and-outputs---Q&A" target="_blank">wiki</a> for the CDS requirement. You may want to check out the toy example in the ./test folder to get familiarized. </p>
				</article>
			</section>

			<section class="container-inputGenomes">
				<div class="title-add-features">
					<p>Data Input</p>
				</div>

				<section class="container-email">
					<p class="session-title">Email Adress</p>
					<input name=emailpangenome id="emailpangenome" class="email" placeholder="Enter email" type="email" onfocus="clearPlaceholder(this)" autocomplete="email"/> 
				</section>
				
				<div id="list-pangenome" class="input-listgenome">
					<div class="session-title">
						<p>Genome data <span class="required">*</span></p>
					</div>

					<div class="box-group">
						<div class="box-input">
							<input type="text" id="fileNameInput-genome" placeholder="FASTA" class="input-text2" readonly disabled>
							<label class="file-input">
							<p class="name-btn">Browser</p>
							<input name="pangenome-file" type="file" class="input-pangenome" accept=".fasta,.fa,.fna,.ffn,.faa,.frn">
							</label>
						</div>

						<div class="box-input">
							<input type="text" id="fileNameInput-cds" placeholder="cds (opcional)" class="input-codingds-text" readonly disabled>
							<label class="file-input">
							<p class="name-btn">Browser</p>
							<input name="codingds-file" type="file" class="input-codingds" accept=".fasta,.fa,.fna,.ffn,.faa,.frn">
							</label>
						</div>
					</div>
				</div>

				<button type="button" id="add-genome" class="add-button">+</button>
			</section>

			<div class="title-add-features">
				<p>Additional features</p>
			</div>
			<div class="pangenome-add-features">
				<section class="container-option">
					<div class="session-title">
						<p> Coding DNA Sequence (cds) <span class="required">*</span></p>
					</div>
					<div class="box-input">
						<input type="text" id="cdspangenomeInput" name="cdspangenome_file" class="input-library" placeholder=".fasta .fa" readonly disabled>
						<label for="cdspangenome" class="file-input">
							<p class="name-btn">Browser</p>
							<input type="file" class="input-data" accept=".fasta,.fa,.fna,.ffn,.faa,.frn" id="cdspangenome" name="cdspangenome_file">
						</label>
					</div>
				</section>
				<section class="container-option">
					<div class="session-title">
						<p> Non-redundant library </p>
					</div>
					<div class="box-input">
						<input type="text" id="nlibraryInput" name="nlibrary_file" class="input-library" readonly disabled>
						<label for="nlibrary" class="file-input">
							<p class="name-btn">Browser</p>
							<input type="file" class="input-data" accept=".fasta,.fa,.fna,.ffn,.faa,.frn" id="nlibrary" name="nlibrary_file">
						</label>
					</div>
				</section>
			</div>

			<div class="container-number-input">
				<section class="container-threadspangenome">
					<p class="session-title"> Threads </p>
					<input type="number" id="threadspangenome" name="threadspangenome" min="10" value="10" step="1">
				</section>
				
				<section class="container-tecopies">
					<label for="tecopies" class="session-title"> Minimum number of full-length TE copies in individual genomes </label>
					<input type="number" id="tecopies" name="tecopies_number" min="3" value="3" title="Enter a value"/>
				</section>
			</div>

			<section class="container-submit-pangenome">
				<button type="button" id="uploadpangenome" class="uploadpangenome"> Submit </button>
			</section>
			
		</section>

		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>

	<main class="results">
		<header class="header">
			<section class="container-header">
				<img class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>
		<section class="container-results">
			<div class="title-input">
				<p>Latest Activities</p>
			</div>
			<article>
				<p> List of the 10 most recent genomic data annotations </p>
			</article>

			<ul class="list-results" id="list-results"></ul>
		</section>
	</main>

	<main class="Help">
		<header class="header">
			<section class="container-header">
				<img class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>
		<section class="documentation-edta">
			<section class="bloco">
				<h3> EDTAGUI documentation - The <span>E</span>xtensive <i><span>D</span>e novo</i> <span>T</span>E <span>A</span>nnotator </h3>
			
				<p> Welcome to the EDTA Documentation </p>
				<p> The EDTA package was designed to filter out false discoveries in raw TE candidates and generate a high-quality non-redundant TE library for whole-genome TE annotations. Selection of initial search programs were based on benckmarkings on the annotation performance using a manually curated TE library in the rice genome.  </p>

				<p> EDTA also features a structured graphical interface that facilitates the annotation of transposable elements in eukaryotic genomes. This interface enables the scientific community to conduct analyses more efficiently and with greater visual accessibility, providing essential support for research focused on the understanding and classification of transposable elements. </p>
			</section>
				
			<section class="bloco">
				<h3> How to use the platform </h3>
				<p> The EDTA interface is divided into four main sections: EDTA, panEDTA, Results, and Help. Below, we explain how to use the first three sections to configure and perform transposable element analysis. </p>
			</section>

			<section class="bloco">
				<h3 style="border-bottom: 0.125rem solid var(--red-01);"> EDTA </h3>
				<p style="margin-bottom: 1rem;"> This section replicates several fields from the EDTA pipeline, organizing them into sub-areas to enhance annotation clarity </p>

				<h4> Data Input </h4>
				<p> Data entry begins with the following fields: </p>
				<p><b>Email Address: </b> To start the annotation process, enter a valid email address. This is a optional field and is used to send notifications about the status of the analysis. While the email facilitates communication, the annotation occurs locally, so keep the system running during the process. </p>
				
				<p><b>Genome Data: </b> Upload the input file containing the complete genomic sequence in FASTA format. Use the "Browse" button to select the file from the local system.</p>
			</section>

			<section class="bloco">
				<h4> Additional Features </h4>
				<p>This subsection provides advanced configuration options for the analysis.</p>

				<p><b>Species Specification to Identify TIR Candidates:</b> Specify the species to optimise the identification of TIR elements. The options include:</p>
				<ul>
					<li><p><b>Others: </b> For species other than rice or maize (default).</p></li>
					<li><p><b>Rice: </b>For rice genomes.</p></li>
					<li><p><b>Maize: </b>For maize genomes. </p></li>
				</ul>

				<p><b>Steps to Be Executed:</b> Select which parts of the pipeline will be executed: </p>
				<ul>
					<li><p><b>All: </b>Runs the entire pipeline (default).</p></li>
					<li><p><b>Filter: </b>Starts from raw TEs to the end.</p></li>
					<li><p><b>Final: </b>Begins with filtered TEs to completion.</p></li>
					<li><p><b>Anno: </b>Conducts genome-wide annotation after building the TE library.</p></li>
				</ul>
				<p>When selecting the Filter, Final, or Anno options, it is necessary to specify the name of the folder where a previous annotation step has already been executed. This folder must be located within the <i>/EDTA/gui/results</i> directory. This ensures that EDTAGUI can access the required files to continue the process, reusing the data already generated and avoiding the repetition of previous steps.</p>

				<p><b>Overwrite:</b> Decide whether existing output data should be overwritten.</p>
				<p><b>Sensitivity:</b> Control the execution of RepeatModeler to identify additional elements.</p>
				<p><b>Annotation:</b> Specify whether the genome-wide annotation of TEs should proceed after building the TE library.</p>
				<p><b>Evaluate:</b> Check if the classification of annotated TEs is consistent. The "Annotation" field must be enabled to use this feature.</p>
				<p><b>Force:</b> If no reliable TE candidates are identified, enabling this option allows the script to continue using a backup rice TE library.</p>

				<p><b>Neutral Mutation Rate:</b> Set the neutral mutation rate for calculating the age of intact LTR elements (default: 1.3e-8 bp per year, based on rice).</p>
				<p><b>Maximum Divergence:</b> Define the maximum acceptable divergence for TE fragments. For highly repetitive genomes, users are encouraged to adjust the parameter (default: 40). </p>
				<p><b>Threads:</b> Determine the number of threads to be used in running the pipeline (default: 4).</p>
			</section>
				
			<section class="bloco">
				<h4> Additional Input File </h4>
				<p>This subsection allows for the addition of optional files to customise the analysis:</p>

				<p><b>Coding DNA Sequence: </b>Select a FASTA file containing the coding sequence (without introns, UTRs, or TEs) of the genome or a close relative. This helps in excluding non-transposable elements.</p>
				<p><b>Curate library: </b>Upload a curated library to maintain consistent TE naming and classification. Only manually validated TEs should be provided. This file is optional.</p>
				<p><b>Exclusion of masked regions: </b>Define regions to be ignored during TE masking. The "Annotation" field must be enabled to use this option.</p>
				<p><b>RepeatModeler library: </b>Upload a classified RepeatModeler library to enhance analysis sensitivity, particularly for LINEs. If not provided, one will be generated automatically.</p>
				<p><b>RepeatMasker library: </b>Provide your own homology-based TE annotation in RepeatMasker .out format. This file will be merged with the structure-based annotation. The "Annotation" field must be enabled.</p>
			</section>

			<section class="bloco">
				<h3 style="border-bottom: 0.125rem solid var(--red-01);"> panEDTA </h3>
				<p style="margin-bottom: 1rem;">This section provides the interface for panEDTA, a new feature of the EDTA pipeline that enables sequential annotation of multiple genomes.</p>

				<h4> Data Input </h4>
				<p>Data input begins with the following fields:</p>
				<p><b>Email Address: </b> To start the annotation process, enter a valid email address. This is a optional field and is used to send notifications about the status of the analysis. While the email facilitates communication, the annotation occurs locally, so keep the system running during the process. </p>
				
				<p><b>Genome Data: </b> Upload the input file containing the complete genome sequence in FASTA format, along with a corresponding CDS file (optional). You must provide at least one genome file. The optional CDS file will only be considered if it is linked to a genome; otherwise, it will be ignored. Use the Browse button to select files from your local system. By clicking the "+" icon, you can add more fields to upload additional genome and CDS files. </p>
			</section>

			<section class="bloco">
				<h4> Additional Features </h4>
				<p>This subsection provides advanced configuration options for the analysis.</p>

				<p><b>Coding DNA Sequence:</b> Required. A coding sequence file in FASTA format. The CDS file provided in this field will be used to fill in any missing CDS files from the Genome Data list. If no CDS files are specified in the genome list, this CDS file will be applied to all genomes.</p>
				<p><b>Non-redundant library:</b> Optional. A manually curated non-redundant library following the RepeatMasker naming convention. </p>
			
				<p><b>Threads:</b> Specify the number of threads to be used when running panEDTA (default: 10).</p>
				<p><b>Minimum number of full-length TE copies in individual genomes:</b> Defines the minimum number of full-length TE copies required in individual genomes for them to be considered as candidate TEs for the pangenome. Lower values are more inclusive, resulting in a larger library, higher sensitivity, but increased inconsistency. Higher values are more stringent, leading to a smaller library, reduced sensitivity, and greater consistency. (default: 3). </p>
				
			</section>
				
			<section class="bloco">
				<h4> Submit and results </h4>
				<p> After completing all necessary fields, click "Submit" to start the analysis. Ensure that all configurations are correct for efficient and accurate processing. </p>

				<p>If you have provided a valid email address, you will receive a notification once the annotation process has started. Upon completion, a second email will be sent, indicating whether the process was successful or if any issues occurred. This final email will include a detailed log file outlining the outcome.</p>

				<p>You can also monitor the progress of the annotation by accessing the "Results" tab. This section displays key information such as:</p>

				<ul>
					<li><p>The name of the generated output file;</p></li>

					<li><p>Start and end timestamps (when available);</p></li>

					<li><p>The current status of the annotation (e.g., in progress, completed, failed);</p></li>

					<li><p>The last 20 lines of the annotation log.</p></li>
				</ul>

				<p><b>Note 1:</b> all results and output files will be stored in the Docker volume you specified as the output directory. Ensure this path is correctly mounted to access the generated data.</p>

				<p><b>Note 2:</b> Errors may occasionally occur during the annotation process, so it is important to pay attention to two key stages:</p>
				<ol>
					<li><p><b>Start of annotation: </b>After uploading a file, ensure that the annotation process has actually begun. Sometimes, the file may not be properly recognised by the system, and in such cases, you will need to re-upload the file and restart the process.</p></li>
					<li><p><b>Completion of annotation: </b>Even if the process runs, the annotation may not be successfully finalised in the Results section. Therefore, it is important to check the following:</p></li>
					<ul style="margin-left: 1.5rem;">
						<li>Look out for any prolonged error messages.</li>
						<li>Confirm whether the message <i>"Evaluation of TE annotation finished! ..."</i> or <i>"panEDTA annotation of genome_${date}.cds.list is finished!"</i> appears — this message indicates that the annotation has been successfully completed.</li>
					</ul>
				</ol>
			</section>
		</section>
		
		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>

	<script src="https://unpkg.com/@phosphor-icons/web"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/main.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/style.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/dark-theme.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/pages.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/panEDTA.js') }}"></script>
</body>
</html>
