<!DOCTYPE html>
<html lang="pt-BR">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link rel="shortcut icon" href="../static/assets/icon.svg" type="image/x-icon" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/home.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/panEDTA.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/results.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/sidebarHelp.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/dark-theme.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/annotep-theme.css') }}" />
    <title> EDTA-GUI </title>
</head>
<body>
	<div class="menuMobile">
		<img src="../static/assets/icon_menu.svg" alt="menu icon" class="menuIcon">
	</div>

	<aside class="aside">
		<ul id="menu">
			<div class="logoedta home" id="Home">
				<img class="icon-menu" src="../static/assets/icon.svg" alt="Logo do edta">
			</div>
			<li class="li open home" id="Home">
				<img class="home-icon" src="../static/assets/home.svg">
				<span class="txt-link"> EDTA-GUI </span>
			</li>
			<li class="li" id="panEDTA">
				<img class="help-icon" src="../static/assets/dna.svg">
				<span class="txt-link"> panEDTA-GUI </span>
			</li>
			<li class="li" id="Results">
				<img class="help-icon" src="../static/assets/box.svg">
				<span class="txt-link"> Results </span>
			</li>
		</ul>
	</aside>

	<div class="cloudImageWrapper">
		<img id="cloudImage" class="replace" src="../static/assets/CloudMoon.svg">
	</div>

	<main class="main sectionsPage">
		<header class="header">
			<section class="container-header">
				<img class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>

		<section class="container-main">
			<article class="description-screen">
				<p> Welcome to EDTA-GUI. This version provides a graphical interface that streamlines your work and research in annotating transposable elements. The performance of the annotation process depends on your machine's specifications, so ensure it remains powered on during execution. For any questions or further information, use the icon<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" style="cursor: text"> &nbsp. </p>
			</article>
		</section>
	
		<section class="container-data-input">
			<div class="title-input">
				<p>Data input</p>
			</div>
			<div class="data-input">
				<section class="container-email">
					<div class="title-help">
						<p class="session-title">Email Adress</p>
						<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar1')">
					</div>
					<input name=email id="email" class="email" placeholder="Enter email" type="email" onfocus="clearPlaceholder(this)"/> 
				</section>
				
				<section class="container-genome">
					<div class="title-help">
						<p class="session-title">Genome data</p>
						<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar1')">
					</div>
					<div  class="input-genome">
						<div class="box-input">
							<input type="text" id="fileNameInput" placeholder="FASTA" class="input-text" readonly disabled>
		
							<label for="inputdata" class="file-input">
								<p class="name-btn">Browser</p>
								<input name="genome-file" type="file" class="input-data" accept=".fasta,.fa,.fna,.ffn,.faa,.frn" id="inputdata">
							</label>
						</div>
					</div>
				</section>
			</div>

			<div id="helpSidebar1" class="help-sidebar hiddenHelp">
				<button class="close-btn" onclick="toggleHelp('helpSidebar1')">
					<img src="../static/assets/close.svg" alt="close" class="close-icon">
				</button>
				<h2> Email Adress </h2>
				<p>To start the annotation process, enter a valid email address. This is a optional field and is used to send notifications about the status of the analysis. While the email facilitates communication, the annotation occurs locally, so keep the system running during the process.</p>

				<p style="margin-top: 1.5rem;"> You can also monitor the progress of the annotation by accessing the "Results" tab. </p>

				<h2 style="margin-top: 1.5rem;"> Genome Data </h2>
				<p>Upload the input file containing the complete genomic sequence in FASTA format. Use the "Browse" button to select the file from the local system.</p>
			</div>

		</section>

		<section class="container-action">
			<div class="title-add-features">
				<p>Additional features</p>
			</div>

			<section style="margin-top: 0.5rem;">
				<div class="title-help">
					<p class="session-title">Mode </p>
					<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar2')">
				</div>
				<div class="box-checkboxs-mode">
					<div class="container-checkbox-mode">
						<input type="checkbox" id="edtagui" name="mode" value="edtagui" class="checkbox-mode" checked onclick="toggleCheckbox('box-checkboxs-mode', this)">
						<label for="edtagui"> EDTA-GUI </label>
					</div>
					<div class="container-checkbox-mode">
						<input type="checkbox" id="annotep" name="mode" value="annotep" class="checkbox-mode" onclick="toggleCheckbox('box-checkboxs-mode', this)">
						<label for="annotep">AnnoTEP (Annotation of TEs in Plants)</label>
					</div>
				</div>

				<div id="helpSidebar2" class="help-sidebar hiddenHelp">
					<button class="close-btn" onclick="toggleHelp('helpSidebar2')">
						<img src="../static/assets/close.svg" alt="close" class="close-icon">
					</button>
					<h2> Mode </h2>
					<p> Specifies which annotation tool you wish to use. The available options are: </p>
					<ul>
						<li><strong>EDTA-GUI</strong> Provides annotation of transposable elements in eukaryotic genomes. Performs <i>de novo</i> detection and general classification of different TE classes. This is the default option for genomes of any eukaryotic organism. (default) </li>
						<li style="margin-top: 1rem;"><strong>AnnoTEP: </strong> Provides specialised annotation of transposable elements in plant genomes. A version of EDTA adapted for high accuracy in identifying all classes, including autonomous and non-autonomous elements, as well as LTR lineages in plant genomes. Produces charts, detailed reports, and phylogenetic trees from the results. </li>
					</ul>
				</div>
			</section>

			<div class="container-add-feature">
				<section>
					<div class="title-help">
						<p class="session-title">Species specification to identify TIR candidates </p>
						<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar3')">
					</div>
					
					<div class="box-checkboxs-tir">
						<div class="container-checkbox-tir">
							<input type="checkbox" id="others" name="tir" value="others" class="checkbox-tir" checked onclick="toggleCheckbox('box-checkboxs-tir', this)">
							<label for="others">others</label>
						</div>
						<div class="container-checkbox-tir">
							<input type="checkbox" id="rice" name="tir" value="rice" class="checkbox-tir" onclick="toggleCheckbox('box-checkboxs-tir', this)">
							<label for="rice">rice</label>
						</div>
						<div class="container-checkbox-tir">
							<input type="checkbox" id="maize" name="tir" value="maize" class="checkbox-tir" onclick="toggleCheckbox('box-checkboxs-tir', this)">
							<label for="maize">maize</label>
						</div>
					</div>

					<div id="helpSidebar3" class="help-sidebar hiddenHelp">
						<button class="close-btn" onclick="toggleHelp('helpSidebar3')">
							<img src="../static/assets/close.svg" alt="close" class="close-icon">
						</button>
						<h2>Species Specification to Identify TIR Candidates </h2>
						<p style="margin-top: 1rem;">The options include: </p>
						<ul>
							<li style="margin-top: 1rem;"><strong>Others:</strong> For species other than rice or maize (default). </li>
							<li style="margin-top: 1rem;"><strong>Rice:</strong> For rice genomes.</li>
							<li style="margin-top: 1rem;"><strong>Maize:</strong> For maize genomes. </li>
						</ul>
					</div>
				</section>
			
				<section>
					<div class="title-help">
						<p class="session-title"> Steps to be executed </p>
						<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar4')">
					</div>
					
					<div class="box-checkboxs-step">
						<div class="container-checkbox-step">
							<input type="checkbox" id="all" name="step" value="all" class="checkbox-step" checked onclick="toggleCheckbox('box-checkboxs-step', this); updateDirectoryInputState()">
							<label for="all">all</label>
						</div>
						<div class="container-checkbox-step">
							<input type="checkbox" id="filter" name="step" value="filter" class="checkbox-step" onclick="toggleCheckbox('box-checkboxs-step', this); updateDirectoryInputState()">
							<label for="filter">filter</label>
						</div>
						<div class="container-checkbox-step">
							<input type="checkbox" id="final" name="step" value="final" class="checkbox-step" onclick="toggleCheckbox('box-checkboxs-step', this); updateDirectoryInputState()">
							<label for="final">final</label>
						</div>
						<div class="container-checkbox-step">
							<input type="checkbox" id="anno" name="step" value="anno" class="checkbox-step" onclick="toggleCheckbox('box-checkboxs-step', this); updateDirectoryInputState()">
							<label for="anno">anno</label>
						</div>
						 <div class="container-directory-name hiddenstep">
							<input type="text" name="directory_name" id="directoryInput" class="directory_name" placeholder="Enter directory name" readonly disabled>
							<label for="directorynameInput" class="file-input">
								<p class="name-btn">Browse</p>
								<input type="file" class="input-data" id="directorynameInput" webkitdirectory directory multiple style="display: none;">
							</label>
						</div>
					</div>

					<div id="helpSidebar4" class="help-sidebar hiddenHelp">
						<button class="close-btn" onclick="toggleHelp('helpSidebar4')">
							<img src="../static/assets/close.svg" alt="close" class="close-icon">
						</button>
						<h2> Steps to Be Executed</h2>
						<p> Select which parts of the pipeline will be executed:</p>
						<ul>
							<li style="margin-top: 1rem;"><strong>All:</strong> Runs the entire pipeline (default). </li>
							<li style="margin-top: 1rem;"><strong>Filter:</strong> Starts from raw TEs to the end. </li>
							<li style="margin-top: 1rem;"><strong>Final:</strong> Begins with filtered TEs to completion. </li>
							<li style="margin-top: 1rem;"><strong>Anno:</strong> Conducts genome-wide annotation after building the TE library. </li>
						</ul>

						<p style="margin-top: 1rem;">When selecting the options "Filter", "Final" or "Anno", a field will appear to choose the folder where a previous annotation step was processed.</p>
						<p>This folder must be located in one of the following directories:</p>
						<ul>
							<li style="margin-top: 1rem;"><strong>Repository installation:</strong> /EDTA/gui/results </li>
							<li style="margin-top: 1rem;"><strong>Docker installation: </strong>  /usr/local/AnnoTEP/gui/results or in the volume configured to map this directory. </li>
						</ul>
					</div>
				</section>
			</div>
			

			<div class="title-help" style="margin-top: 1.5rem; margin-bottom: 0;">
					<p class="session-title"> TE Analysis Parameters </p>
					<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar5')">
		   </div>
		   <section class="container-switch">
				<table class="table-switch">
					<tr>
						<td><p class="session-title">Overwrite</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch1" name="overwrite"/>
								<label for="switch1">Toggle</label>
							</div>
						</td>
						<td><p id="status-text-switch1"> Deactivated </p></td>
					</tr>
					<tr>
						<td><p class="session-title">Sensitivity</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch2" name="sensitivity"/>
								<label for="switch2">Toggle</label>
							</div>
						</td>
						<td><p id="status-text-switch2"> Deactivated </p></td>
					</tr>
					<tr>
						<td><p class="session-title">Annotation</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch3" name="annotation"/>
								<label for="switch3">Toggle</label>
							</div>
						</td>
						<td><p id="status-text-switch3"> Deactivated</p></td>
					</tr>
				</table>

				<table class="table-switch">
					<tr>
						<td><p class="session-title">Evaluate</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch4" name="evaluate"/>
								<label for="switch4">Toggle</label>   
							</div>
						</td>
						<td><p id="status-text-switch4">Deactivated</p></td>
					</tr>
					<tr>
						<td><p class="session-title">Force</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch5" name="force"/>
								<label for="switch5">Toggle</label>
							</div>
						</td>
						<td><p id="status-text-switch5">Deactivated</p></td>
					</tr>
					<tr class="hidden">
						<td><p class="session-title">TIR filter</p></td>
						<td class="button-switch">
							<div class="switch">
								<input type="checkbox" id="switch6" name="tirfilter"/>
								<label for="switch6">Toggle</label>   
							</div>
						</td>
						<td><p id="status-text-switch6">Deactivated</p></td>
					</tr>
				</table>

				<table class="table-switch">
					<tr class="hidden">
						<td><p class="session-title">Annot. type</p></td>
						<td class="button-switch" style="padding-left: 0.5rem;">
							<div class="switch">
								<input type="checkbox" id="switch7" name="annottype"/>
								<label for="switch7">Toggle</label>   
							</div>
						</td>
						<td><p id="status-text-switch7">Deactivated</p></td>
					</tr>
				</table>
		   </section>

			<div class="container-number-input">
				<section class="container-mutation-rate">
					<label for="mutation-rate" class="session-title"> Neutral mutation rate </label>
					<input type="text" id="mutation-rate" name="mutation_rate" placeholder="1.3e-8" title="Enter a value in scientific notation, for example, 1.3e-8"/>
				</section>
	
				<section class="container-maxdiv">
					<label for="maximun-divergence" class="session-title"> Maximum divergence </label>
					<input type="number" id="maximun-divergence" name="max_divergence" value="40" min="0" max="100" title="Enter a value from 0 to 100"/>
				</section>
	
				<section class="container-threads">
					<p class="session-title"> Threads </p>
					<input type="number" id="threads" name="threads" min="10" value="10" step="1	">
				</section>
			</div>
			 
		</section>

			<section class="container-action">
				<div class="title-add-features">
					<p>Additional input files</p>
					<span class="not-required"> (Field not required) </span>
				</div>

				<div class="container-add-input">
					<section class="container-genome">
						<div class="title-help">
							<p class="session-title">  Coding DNA Sequence (cds) </p>
							<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar6')">
						</div>
						<div class="box-input">
							<input type="text" id="cdsInput" name="cds_file" class="input-library" placeholder=".fasta .fa" readonly disabled>
							<label for="cds" class="file-input">
								<p class="name-btn">Browser</p>
								<input type="file" class="input-data" accept="fasta,.fa,.fna,.ffn,.faa,.frn" id="cds">
							</label>
						</div>
					</section>

					<section class="container-genome">
						<div class="title-help">
							<p class="session-title">  Curate library (curatedlib) </p>
							<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar6')">
						</div>
						<div class="box-input">
							<input type="text" id="curateLibInput" name="curate_lib" class="input-library"  readonly disabled>
							<label for="curatelib" class="file-input">
								<p class="name-btn">Browser</p>
								<input type="file" class="input-data" id="curatelib">
							</label>
						</div>
					</section>

					<section class="container-genome">
						<div class="title-help">
							<p class="session-title">  Exclusion of masked regions (exclude) </p>
							<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar6')">
						</div>
						<div class="box-input">
							<input type="text" id="maskedRegionsInput" name="masked_regions" class="input-library" placeholder=".bed" readonly disabled>
		
							<label for="exclude" class="file-input">
								<p class="name-btn disabled" id="browseExcludeButton">Browser</p>
								  <input type="file" class="input-data" accept=".bed" id="exclude">
							</label>
						</div>
					</section>
				</div>

				<div class="container-add-input">
					<section class="container-genome">
						<div class="title-help">
							<p class="session-title">  RepeatModeler library (rmlib) </p>
							<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar6')">
						</div>
						<div class="box-input">
							<input type="text" id="rmLibInput" name="rm_lib" class="input-library" readonly disabled>
		
							<label for="rmlib" class="file-input">
								<p class="name-btn">Browser</p>
								  <input type="file" class="input-data" id="rmlib">
							</label>
						</div>
					</section>

					<section class="container-genome">
						<div class="title-help">
							<p class="session-title">  RepeatMasker library (rmout) </p>
							<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar6')">
						</div>
						<div class="box-input">
							<input type="text" id="rmoutInput" name="rm_out" class="input-library" placeholder=".out" readonly disabled>
							<label for="rmout" class="file-input">
								<p class="name-btn disabled" id="browseButton">Browser</p>
								<input type="file" class="input-data" accept=".out" id="rmout" disabled>
							</label>
						</div>
					</section>
				</div>

				<div id="helpSidebar5" class="help-sidebar hiddenHelp">
					<button class="close-btn" onclick="toggleHelp('helpSidebar5')">
						<img src="../static/assets/close.svg" alt="close" class="close-icon">
					</button>
					<h2>TE Analysis Parameters</h2>
					<h3>EDTA-GUI mode</h3>
						<ul>
							<li><strong>Overwrite:</strong> Decide whether existing output data should be overwritten. </li>
							<li><strong>Sensitivity:</strong> Control the execution of RepeatModeler to identify additional elements. </li>
							<li><strong>Annotation:</strong> Specify whether the genome-wide annotation of TEs should proceed after building the TE library.</li>
							<li><strong>Evaluate:</strong> Check if the classification of annotated TEs is consistent. The "Annotation" field must be enabled to use this feature.</li>
							<li><strong>Force:</strong> If no reliable TE candidates are identified, enabling this option allows the script to continue using a backup rice TE library.</li>
						</ul>
						<h3> AnnoTEP mode (Additional feature)</h3>
						<ul>
							<li><strong>TIR filter:</strong> Filter TIRs without annotated domains. Enabling this filter can substantially reduce false positives, but may also result in the loss of some true positives (false negatives).</li>
							<li><strong>Annot. type:</strong>  Specify whether to annotate the genome using a RepeatMasker-based librar. Enabling this option may negatively affect the filtering step and compromise benchmark results. </li>
						</ul>
						<h3> </h3>
						<ul>
							<li><strong>Neutral Mutation Rate:</strong> Set the neutral mutation rate for calculating the age of intact LTR elements (default: 1.3e-8 bp per year, based on rice).</li>
							<li><strong>Maximum Divergence:</strong>  Define the maximum acceptable divergence for TE fragments. For highly repetitive genomes, users are encouraged to adjust the parameter (default: 40). </li>
							<li><strong>Threads:</strong>  Determine the number of threads to be used in running the pipeline (default: 10). </li>
						</ul>
				</div>

				<div id="helpSidebar6" class="help-sidebar hiddenHelp">
					<button class="close-btn" onclick="toggleHelp('helpSidebar6')">
						<img src="../static/assets/close.svg" alt="close" class="close-icon">
					</button>
					<h2>Additional Input File</h2>
					<p style="margin-top: 1rem;">This subsection allows for the addition of optional files to customise the analysis:</p>
					<ul>
						<li style="margin-top: 1rem;"><strong>Coding DNA Sequence:</strong> Select a FASTA file containing the coding sequence (without introns, UTRs, or TEs) of the genome or a close relative. This helps in excluding non-transposable elements. </li>
						<li style="margin-top: 1rem;"><strong>Curate library:</strong> Upload a curated library to maintain consistent TE naming and classification. Only manually validated TEs should be provided. This file is optional. </li>
						<li style="margin-top: 1rem;"><strong>Exclusion of masked regions:</strong> Define regions to be ignored during TE masking. The "Annotation" field must be enabled to use this option.</li>
						<li style="margin-top: 1rem;"><strong>RepeatModeler library:</strong> Upload a classified RepeatModeler library to enhance analysis sensitivity, particularly for LINEs. If not provided, one will be generated automatically.</li>
						<li style="margin-top: 1rem;"><strong>RepeatMasker library:</strong>Provide your own homology-based TE annotation in RepeatMasker .out format. This file will be merged with the structure-based annotation. The "Annotation" field must be enabled. </li>
					</ul>
				</div>
			</section>
	
		<section class="container-submit">
			<button type="button" id="uploaddata" class="uploaddata"> Submit </button>
		</section>

		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>

	<main class="panEDTA">
		<header class="header">
			<section class="container-header">
				<img class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>
		<section class="container-panGenome">
			<div class="title-input">
				<p>panEDTA-GUI</p>
			</div>
			<section class="container-main">
				<article class="description-screen">
					<p>This is the serial version of panEDTA. Each genome will be annotated sequentially and then combined with the panEDTA functionality. Existing EDTA annotation of genomes (EDTA run with --anno 1) will be recognized and reused. A way to acclerate the pan-genome annotation is to execute EDTA-GUI annotation of each genomes separately and in parallel, then execute panEDTA-GUI to finish the remaining of the runs. To help filtering out gene-related sequences, at least one CDS file is required. Please read <a href="https://github.com/oushujun/EDTA/wiki/Making-sense-of-EDTA-usage-and-outputs---Q&A" target="_blank">wiki</a> for the CDS requirement. You may want to check out the toy example in the ./test folder to get familiarized. </p>
				</article>
			</section>

			<section class="container-inputGenomes">
				<div class="title-add-features">
					<p>Data Input</p>
				</div>

				<section class="container-email">
					<div class="title-help">
						<p class="session-title">Email Adress</p>
						<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar7')">
					</div>
					<input name=emailpangenome id="emailpangenome" class="email" placeholder="Enter email" type="email" onfocus="clearPlaceholder(this)" autocomplete="email"/> 
				</section>
				
				<div id="list-pangenome" class="input-listgenome">
					<div class="title-help">
						<div class="session-title">
							<p>Genome data <span class="required">*</span></p>
						</div>
						<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar7')">
					</div>

					<div class="box-group">
						<div class="box-input">
							<input type="text" id="fileNameInput-genome" placeholder="FASTA" class="input-text2" readonly disabled>
							<label class="file-input">
							<p class="name-btn">Browser</p>
							<input name="pangenome-file" type="file" class="input-pangenome" accept=".fasta,.fa,.fna,.ffn,.faa,.frn">
							</label>
						</div>

						<div class="box-input">
							<input type="text" id="fileNameInput-cds" placeholder="cds (opcional)" class="input-codingds-text" readonly disabled>
							<label class="file-input">
							<p class="name-btn">Browser</p>
							<input name="codingds-file" type="file" class="input-codingds" accept=".fasta,.fa,.fna,.ffn,.faa,.frn">
							</label>
						</div>
					</div>
				</div>

				<button type="button" id="add-genome" class="add-button">+</button>

				<div id="helpSidebar7" class="help-sidebar hiddenHelp">
					<button class="close-btn" onclick="toggleHelp('helpSidebar7')">
						<img src="../static/assets/close.svg" alt="close" class="close-icon">
					</button>
					<h2> Email Adress </h2>
					<p>To start the annotation process, enter a valid email address. This is a optional field and is used to send notifications about the status of the analysis. While the email facilitates communication, the annotation occurs locally, so keep the system running during the process.</p>

					<p style="margin-top: 1.5rem;"> You can also monitor the progress of the annotation by accessing the "Results" tab. </p>

					<h2 style="margin-top: 1.5rem;"> Genome Data </h2>
					<p>Upload the input file containing the complete genome sequence in FASTA format, along with a corresponding CDS file (optional). You must provide at least one genome file. The optional CDS file will only be considered if it is linked to a genome; otherwise, it will be ignored. Use the Browse button to select files from your local system. By clicking the "+" icon, you can add more fields to upload additional genome and CDS files. By clicking on the "-"" icon, you can remove the added field.</p>
				</div>
			</section>

			<div class="title-add-features">
				<p>Additional features</p>
			</div>
			<div class="pangenome-add-features">
				<section class="container-option">
					<div class="title-help">
						<div class="session-title">
							<p>Coding DNA Sequence (cds)  <span class="required">*</span></p>
						</div>
						<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar8')">
					</div>
					<div class="box-input">
						<input type="text" id="cdspangenomeInput" name="cdspangenome_file" class="input-library" placeholder=".fasta .fa" readonly disabled>
						<label for="cdspangenome" class="file-input">
							<p class="name-btn">Browser</p>
							<input type="file" class="input-data" accept=".fasta,.fa,.fna,.ffn,.faa,.frn" id="cdspangenome" name="cdspangenome_file">
						</label>
					</div>
				</section>
				<section class="container-option">
					<div class="title-help">
						<p class="session-title">Non-redundant library</p>
						<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar8')">
					</div>
					<div class="box-input">
						<input type="text" id="nlibraryInput" name="nlibrary_file" class="input-library" readonly disabled>
						<label for="nlibrary" class="file-input">
							<p class="name-btn">Browser</p>
							<input type="file" class="input-data" accept=".fasta,.fa,.fna,.ffn,.faa,.frn" id="nlibrary" name="nlibrary_file">
						</label>
					</div>
				</section>
			</div>

			<div class="container-number-input">
				<section class="container-threadspangenome">
					<p class="session-title"> Threads </p>
					<input type="number" id="threadspangenome" name="threadspangenome" min="10" value="10" step="1">
				</section>
				
				<section class="container-tecopies">
					<label for="tecopies" class="session-title"> Minimum number of full-length TE copies in individual genomes </label>
					<input type="number" id="tecopies" name="tecopies_number" min="3" value="3" title="Enter a value"/>
				</section>
			</div>

			<section class="container-submit-pangenome">
				<button type="button" id="uploadpangenome" class="uploadpangenome"> Submit </button>
			</section>

			<div id="helpSidebar8" class="help-sidebar hiddenHelp">
				<button class="close-btn" onclick="toggleHelp('helpSidebar8')">
					<img src="../static/assets/close.svg" alt="close" class="close-icon">
				</button>
				<h2> Additional Features </h2>
				<p> This subsection provides advanced configuration options for the analysis.</p>
				<ul>
					<li style="margin-top: 1rem;"><strong>Coding DNA Sequence:</strong> Required. A coding sequence file in FASTA format. The CDS file provided in this field will be used to fill in any missing CDS files from the Genome Data list. If no CDS files are specified in the genome list, this CDS file will be applied to all genomes. </li>
					<li style="margin-top: 1rem;"><strong>Non-redundant library:</strong> SOptional. A manually curated non-redundant library following the RepeatMasker naming convention. </li>
					<li style="margin-top: 1rem;"><strong>Threads:</strong> Specify the number of threads to be used when running panEDTA-GUI (default: 10). </li>
					<li style="margin-top: 1rem;"><strong>Minimum number of full-length TE copies in individual genomes:</strong> Defines the minimum number of full-length TE copies required in individual genomes for them to be considered as candidate TEs for the pangenome. Lower values are more inclusive, resulting in a larger library, higher sensitivity, but increased inconsistency. Higher values are more stringent, leading to a smaller library, reduced sensitivity, and greater consistency. (default: 3). </li>
				</ul>
			</div>	
		</section>

		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>

	<main class="results">
		<header class="header">
			<section class="container-header">
				<img class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>
		<section class="container-results">
			<div class="title-input">
				<p>Latest Activities</p>
			</div>

			<div class="title-help">
				<article>
					<p> List of the 10 most recent genomic data annotations </p>
				</article>
				<img src="../static/assets/question-red.svg" alt="help icon" class="question-icon" onclick="toggleHelp('helpSidebar9')">
			</div>
			
			<ul class="list-results" id="list-results"></ul>

			<div id="helpSidebar9" class="help-sidebar hiddenHelp">
				<button class="close-btn" onclick="toggleHelp('helpSidebar9')">
					<img src="../static/assets/close.svg" alt="close" class="close-icon">
				</button>
				<h2> Latest Activities </h2>
				<p> This section displays key information such as: </p>
				<ul>
					<li style="margin-top: 1rem;">The name of the generated output file;</li>
					<li style="margin-top: 1rem;">Start and end timestamps (when available);</li>
					<li style="margin-top: 1rem;">The current status of the annotation (e.g., in progress, completed, failed);</li>
					<li style="margin-top: 1rem;">The last 20 lines of the annotation log.</li>
				</ul>

				<h3>Note 1</h3>
				<p> All results and output files will be stored in the Docker volume you specified as the output directory. Ensure this path is correctly mounted to access the generated data.</p>

				<h3>Note 2 </h3>
				<p>Errors may occasionally occur during the annotation process, so it is important to pay attention to two key stages:</p>
				<ul>
					<li style="margin-top: 1rem;">Look out for any prolonged error messages.</li>
					<li style="margin-top: 1rem;"> <strong>Using EDTA-GUI:</strong> Confirm whether the message "Evaluation of TE annotation finished! ..." or "panEDTA annotation of genome_${date}.cds.list is finished!" appears — this message indicates that the annotation has been successfully completed.</li>
					<li style="margin-top: 1rem;"><strong>Using AnnoTEP:</strong> Confirm whether the message "The generation of charts and reports has been completed" appears — this message indicates that the annotation has been successfully completed.</li>
				</ul>
			</div>
		</section>
	</main>

	<script src="https://unpkg.com/@phosphor-icons/web"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/main.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/style.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/dark-theme.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/pages.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/panEDTA.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/annotep-theme.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/results.js') }}"></script>
</body>
</html>
